Implement chunk rotation

jiekang opened the pull request:
This implements the rest of the chunk rotation process.

Safepoint:
Uninterruptibly writes all buffers to disk, notifies Java thread local event writers to restart their writes and changes the epoch

Post-safepoint:
Writes checkpoint event, metadata event, patches chunk header and closes the chunk file


The patch relies on Christian's work to add support for uninterruptibly writing big endian data to disk.

christianhaeubl committed:
Added support for writing big endian data to files.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -57,27 +61,32 @@
 
     private String filename;
-    private RandomAccessFile file;
+    private RawFileOperationSupport fileOperationSupport;
```
`RawFileOperationSupport.bigEndian()` is annotated with `@Fold`, so it is better performance-wise to avoid the field and to just call `RawFileOperationSupport.bigEndian()` instead (it is guaranteed that this will fold to a constant). To make the code a bit shorter, you could add a helper method to `JfrChunkWriter` that is annotated with `@Fold` and that calls `RawFileOperationSupport.bigEndian()`.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -94,7 +103,7 @@ public void unlock() {
     @Override
     public boolean hasOpenFile() {
-        return file != null;
+        return filename != null;
```
It is possible that the filename was already set to a non-null value but that the file was not opened yet. I think that this method needs to check if there is a valid file descriptor (`RawFileOperationSupport.isValid(...)`). 

The default value for the file descriptor (i.e., 0) is a valid file descriptor on Posix but not on Windows. So, `RawFileOperationSupport.isValid(...)` would wrongly return true for the default value value on Posix at the moment. To avoid platform-dependent issues, it might be better to ensure that the default value is not a valid file descriptor on all platforms. So, for now, I would suggest to change `PosixRawFileOperationSupport.isValid(...)` so that it checks for `posixFd > 0` (`stdin` won't be needed at the moment anyways).

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -126,17 +135,34 @@ public boolean openFile(String outputFile) {
+    }
+
+    @Uninterruptible(reason = "Called by uninterruptible code")
+    public boolean writeAtSafepoint(JfrBuffer buffer) {
```
This method is the same as the method above except for the assertion. I would suggest to merge those two method and to annotate the resulting method with `@Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)`

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -126,17 +135,34 @@ public boolean openFile(String outputFile) {
+            return false;
+        }
+        JfrBufferAccess.increaseTop(buffer, unflushedSize);
+        return fileOperationSupport.position(fd).rawValue() > notificationThreshold;
```
Use `return fileOperationSupport.position(fd).greaterThan(notificationThreshold)` instead.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -371,43 +412,61 @@ private static int makePaddedInt(long sizeWritten) {
         @Uninterruptible(reason = "Prevent pollution of the current thread's thread local JFR buffer.")
-        private void changeEpoch() {
+        private boolean changeEpoch() {
             // TODO: We need to ensure that all JFR events that are triggered by the current thread
```
I think you can remove the TODO and convert the comment into a JavaDoc comment that is placed on the method.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrBuffer.java:
```diff
@@ -56,6 +56,12 @@ static int offsetOfPos() {
     }
 
+    @RawField
+    Pointer getTop();
```
It don't think that we need a top pointer at the moment:
- When the VM is at a safepoint, it is guaranteed that all `JfrNativeEventWriter`s finished execution.
- All Java `EventWriter`s won't increase pos before the event write is finished. If a epoch change happens while writing the data, they will be notified and will retry the whole write operation.

jiekang replied:
I think the top pointer is needed for the implementation in this PR because of the EventWriter view that is modified when flushing Java buffers to global buffer but isn't (currently) modified when writing Java buffers to disk at safepoint.

The buffers don't get reinitialized when writing to disk. If they did, we would also need to execute an update of the EventWriter similar to:
https://github.com/native-image-jfr/graal/blob/jfr/substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/SubstrateJVM.java#L340

I think it makes sense to handle it this way for the long term because
* The "lazy" reinitialize shouldn't increase cost for threads writing many events and will reduce cost for threads that write few events (less than size of thread local buffer)
* The work done in the uninterruptible safepoint operation should be as minimal as possible
* It uses the same infrastructure and pattern as concurrent writes of buffers to disk so less will need to be changed when making that optimization

christianhaeubl replied:
Yes, we are definitely going to need it in the long run. After giving this some more thought, I think that it is indeed better to add that now (every other solution would probably be more effort).

However, this has a couple implications on correctness that need to be fixed. E.g. all code that accesses or modifies the top pointer must be annotated with something like `@Uninterruptible(reason = "Prevent safepoints as those could change the top pointer")`. And, all those methods must no longer use `mayBeInlined = true` because the value could change at a safepoint (inlining may break the uninterruptibility). This is necessary to prevent that cases where we would continue using an old top pointer even though it changed at a safepoint. Pretty much the same is necessary for all callers of those methods.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -57,27 +61,32 @@
-    private static final int JFR_VERSION_MAJOR = 2;
-    private static final int JFR_VERSION_MINOR = 0;
+    private static final short JFR_VERSION_MAJOR = 2;
+    private static final short JFR_VERSION_MINOR = 0;
```
In the initial RedHat PR, you used version 2.1. Is there any relevant difference between 2.0 and 2.1?

jiekang replied:
Yes. My understanding from reading the source of 11/latest is that 2.1 introduced to help support jfr streaming (I think) and includes a chunk header checkpoint at the end of every chunk. There is an extra `metadata.xml` entry for type `ChunkHeader` that isn't in 11. You can see more about it here:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/jfr/recorder/repository/jfrChunkWriter.cpp#L192

In our PR we tried to implement a specific commit in latest instead of in 11 in order to best support streaming and future features so we were working to also have this chunk header checkpoint. I got stuck on this `metadata.xml` being different between 11 and latest, so was looking at having a substratevm specific `metadata.xml` to decouple ourselves from the jdk implementation.

I think in the long term I'd like to have a decoupled `metadata.xml` to also reduce the size of the metadata and constant pool events. (e.g. metadata event is currently 617544 bytes for test code that emits a single custom event, when it really doesn't need to be). But 1. It's on the backburner in terms of priorty and 2. I'm not strongly for/against anything here. It's a tradeoff to think about.

christianhaeubl replied:
What do you mean with decoupling the `metadata.xml`? I think that we need to use the same `metadata.xml` at run-time and during the image build. Otherwise, we risk that data that is generated during the image build doesn't match the run-time data.

jiekang replied:
Ah; I mean to have a SubstrateVM specific metadata (xml file or maybe created during the build process) that isn't a carbon copy from the JDK that is being used to build the native image. My hope is to be able to add/remove metadata as we see fit. One to potentially support jfr v2.1 with the `ChunkHeader` type and two to drop all of the HotSpot specific metadata like Shenandoah/Z1 events, etc.

rkennke replied:
I am not sure about the metadata.xml discussion here. I thought much of the point of what I've been doing was to make this unnecessary? I.e. let the host VM use its metadata and use that everywhere, even at run-time? We are only asking for trouble if we were to use our own metadata.xml instead IMO. If/when we need to trim the metadata to not include stuff that is specific to OpenJDK (e.g. G1 Shenandoah ZGC related stuff) then we can come up with a filter, I think.

jiekang replied:
Right. I think overall it makes sense to start with the same metadata as the host VM but I'd definitely want to trim it or add to it. 

For me it was a question of where we want work to be done to manage changing of the metadata between HotSpot versions of event types that we use. The fix for the Thread event sort of illustrates the problem. If we have a native event that has 3 fields in JDK 11, but then has 2 or 4 fields in JDK 17, do we want to use the metadata from HotSpot and then have an abstract event which emits differently for JDK 11 and for JDK 17? Or can we lock the SubstrateVM version of the event to a specific implementation and leave it at that? Event abstraction isn't too bad of a long-term maintenance burden but it gets more complicated when we look at e.g. JFR streaming support in 17 but not 11 and how we want to handle it in SubstrateVM. I don't think it's viable to have infrastructure abstraction in SubstrateVM that supports streaming on 17 but not on 11. This is one case where it also touches metadata that exists in 17 but not in 11. 

Who knows what other major changes will be made to JFR in the future. A "decoupled" SubstrateVM implementation would, to me, seek to meet a JFR spec (I wish one existed in the public) and be free to change itself as needed and on it's own timeline to support things. This would include the metadata it chooses to emit. I realize though I haven't been thinking too much about the end result's use in JMC. I suppose it's ideal if users can take the JFR files from SubstrateVM and use all the same charts they use in JMC for their application on JDK 11 or 17, or whatever.

P.s. I realize streaming was earlier than 17; just using 11 and 17 as example versions!

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -126,17 +135,34 @@ public boolean openFile(String outputFile) {
-        } catch (IOException e) {
-            Logger.log(JFR_SYSTEM, ERROR, "Error while writing file " + filename + ": " + e.getMessage());
+        assert lock.isHeldByCurrentThread()  || VMOperationControl.isDedicatedVMOperationThread() && lock.isLocked();
+        UnsignedWord unflushedSize = JfrBufferAccess.getUnflushedSize(buffer);
```
Please add an assertion that the buffer was acquired or that we are at a safepoint.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -150,50 +176,61 @@ public boolean write(JfrBuffer buffer) {
         op.enqueue();
+
+        try {
+            SignedWord constantPoolPosition = writeCheckpointEvent(repositories);
```
Please document why it is safe to do those operations outside the VM operation.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -150,50 +176,61 @@ public boolean write(JfrBuffer buffer) {
+    private SignedWord writeCheckpointEvent(JfrRepository[] repositories) throws IOException {
         JfrSerializer[] serializers = JfrSerializerSupport.get().getSerializers();
 
         // TODO: Write the global buffers of the previous epoch to disk. Assert that none of the
```
This comment can be removed now.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -231,40 +268,36 @@ private void writeRepositories(JfrRepository[] constantPools) throws IOException
-            Logger.log(JFR_SYSTEM, ERROR, "Could not check file size to determine chunk rotation: " + ex.getMessage());
-            return false;
-        }
+        return filename != null && fileOperationSupport.size(fd).rawValue() > notificationThreshold;
```
See similar comment above. In general, we only use `rawValue()` if something really can't be expressed in a different way.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -231,40 +268,36 @@ private void writeRepositories(JfrRepository[] constantPools) throws IOException
+
+    public void endEvent(SignedWord start) throws IOException {
+        SignedWord end = fileOperationSupport.position(fd);
+        long writtenBytes = end.rawValue() - start.rawValue();
```
`SignedWord writtenBytes = end.subtract(start);` and `makePaddedInt(writtenBytes.rawValue())`

We probably will need at least need some kind of error handling for certain file operations.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -291,54 +324,62 @@ public void writeCompressedLong(long value) throws IOException {
+        fileOperationSupport.writeByte(fd, (byte) (v >>> 7)); // 56-63, last byte as is.
+    }
+
+    public void close() throws IOException {
```
Does not throw an `IOException`. Is this method used?

jiekang replied:
Unused and removed; I think it crept in when I was cherry-picking and rebasing other branches, sorry.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -371,43 +412,61 @@ private static int makePaddedInt(long sizeWritten) {
+            boolean shouldNotify = changeEpoch();
+            if (shouldNotify) {
+                //Checkstyle: stop
+                synchronized (Target_jdk_jfr_internal_JVM.FILE_DELTA_CHANGE) {
```
This can result in a deadlock (there is no guarantee that `Target_jdk_jfr_internal_JVM.FILE_DELTA_CHANGE` is not locked by some other thread). So, the notify needs to be done outside the VM operation. It should be sufficient to do the check at the end of `closeFile()` as no new file can be opened anyways before the old file is closed.

jiekang replied:
I've moved this out as suggested using a boolean field in the class.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -371,43 +412,61 @@ private static int makePaddedInt(long sizeWritten) {
+            JfrBuffers buffers = globalMemory.getBuffers();
+            for (int i = 0; i < globalMemory.getBufferCount(); i++) {
+                JfrBuffer buffer = buffers.addressOf(i).read();
+                if (JfrBufferAccess.acquire(buffer)) {
```
None of the global buffers may be acquired by another thread (otherwise, this would cause major issues as that might not get written into the correct chunk). So, it should be fine to replace `if (JfrBufferAccess.acquire(buffer))` with an assertion that the buffer is not acquired by any thread (and because we are at a safepoint in uninterruptible code, it is also unnecessary to acquire the buffer here).

jiekang replied:
Edit: replied to wrong comment.

This has been fixed.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrGlobalMemory.java:
```diff
@@ -73,11 +73,13 @@ public void teardown() {
         }
     }
 
+    @Uninterruptible(reason = "May be called by uninterruptible code", mayBeInlined = true)
```
Use `@Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)` instead (we use that specific wording for the reason throughout the project). Same applies to the method below.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrThreadLocal.java:
```diff
@@ -205,6 +205,23 @@ public JfrBuffer getNativeBuffer() {
     }
 
+    @Uninterruptible(reason = "Accesses a JFR buffer.")
+    public static JfrBuffer getJavaBuffer(IsolateThread thread) {
```
Please assert that this method is only executed in a safepoint. `callerMustBe = true` can be added to the `@Uninterruptible` annotation.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrThreadLocal.java:
```diff
@@ -205,6 +205,23 @@ public JfrBuffer getNativeBuffer() {
+
+    @Uninterruptible(reason = "Accesses a JFR buffer.", callerMustBe = true)
+    public static JfrBuffer getNativeBuffer(IsolateThread thread) {
+        return nativeBuffer.get(thread);
```
Please assert that this method is only executed in a safepoint.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrThreadLocal.java:
```diff
@@ -205,6 +205,23 @@ public JfrBuffer getNativeBuffer() {
+        return nativeBuffer.get(thread);
+    }
+
+    @Uninterruptible(reason = "Called by uninterruptible code.", callerMustBe = true)
```
Use `@Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)` instead.

jiekang commented:
There is a curious bug with the write of the native thread local buffers to disk where the jdk 11 `jfr print` command fails to read the result, while the jdk latest binary (from openjdk/jdk) succeeds.

Failure is:

```
$ jfr print tcec-broken.jfr
jfr print: could not read recording at /home/jkang/Work/code/src/github.com/native-image-jfr/graal/substratevm/tcec-broken.jfr. Event can't have zero size
```

When manually examining bytes with my reader tool, I don't see any problematic data though. Also the summary command works for both versions of the jfr tool. When the write of native buffers to disk is commented out, the result is readable by the jdk 11 `jfr print` command. I'm reviewing the JfrNativeEventWriter code now and will hope to resolve this bug with further commits to this PR.

christianhaeubl commented on substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixRawFileOperationSupport.java:
```diff
@@ -55,10 +65,11 @@ public RawFileDescriptor open(File file, FileAccessMode mode) {
     public boolean isValid(RawFileDescriptor fd) {
         int posixFd = getPosixFileDescriptor(fd);
-        return posixFd >= 0;
+        return posixFd > 0;
```
Please add a comment that this is done to ensure that the default value 0 is invalid on all platforms.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -125,75 +134,92 @@ public boolean openFile(String outputFile) {
-     * infrastructure is epoch-based. So, we can uninterruptibly switch to a new epoch before we
-     * start writing out the data of the old epoch.
+     * includes allocation and GC events. Therefore, it is necessary that we switch
+     * to a new epoch in an uninterruptible safepoint
```
"... in uninterruptible code at a safepoint."

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -125,75 +134,92 @@ public boolean openFile(String outputFile) {
+            Logger.log(JFR_SYSTEM, ERROR, "Error while writing file " + filename + ": " + e.getMessage());
+        }
+
+        filename = null;
```
The file descriptor field needs to be set to a null pointer as well (`WordFactory.nullPointer()`)

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -125,75 +134,92 @@ public boolean openFile(String outputFile) {
         }
+
+        boolean success = getFileSupport().write(fd, buffer.getTop(), unflushedSize);
+        if (!success) {
```
This error handling is problematic. If we are at a safepoint and the write fails, we wouldn't increase the top pointer. So, we would still have unflushed data from the previous epoch in the buffer.

jiekang replied:
I've adjusted this to increase regardless so failures will *just* be lost data and added a todo about it for the time being

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -125,75 +134,92 @@ public boolean openFile(String outputFile) {
+        if (shouldNotify) {
+            shouldNotify = false;
+            //Checkstyle: stop
+            synchronized (Target_jdk_jfr_internal_JVM.FILE_DELTA_CHANGE) {
```
I am wondering if there are still deadlocks possible, i.e., is there any (JDK) code that could acquire `JfrChunkWriter.lock` while holding `Target_jdk_jfr_internal_JVM.FILE_DELTA_CHANGE`?

Actually, it might be possible to completely remove this code. We already closed the file, so there shouldn't be any need to do the notify. At least I don't think that HotSpot does notify in this particular situation.

jiekang replied:
I reviewed the notify mechanism; I think Victor is looking at revising it to be similar to HotSpot and you're right, the notify shouldn't be here, so it's removed now.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -231,40 +257,36 @@ private void writeRepositories(JfrRepository[] constantPools) throws IOException
-            Logger.log(JFR_SYSTEM, ERROR, "Could not check file size to determine chunk rotation: " + ex.getMessage());
-            return false;
-        }
+        return filename != null && getFileSupport().size(fd).greaterThan(WordFactory.signed(notificationThreshold));
```
This should check if the file descriptor is valid (instead of checking for the filename).

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -371,43 +398,52 @@ private static int makePaddedInt(long sizeWritten) {
+         * - Flush all buffers (native, Java and global) to disk
+         * - Set all Java EventWriter.notified values
+         * - Change the epoch
+         * @return Whether to notify of file changes
```
Is a void method.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -371,43 +398,52 @@ private static int makePaddedInt(long sizeWritten) {
-            // - Flush all thread-local buffers (native & Java) to global JFR memory.
-            // - Set all Java EventWriter.notified values
-            // - Change the epoch.
+            for (IsolateThread thread = VMThreads.firstThread(); thread.isNonNull(); thread = VMThreads.nextThread(thread)) {
```
Please add a comment that it is on purpose that thread-local buffers are not reinitialized.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrRecorderThread.java:
```diff
@@ -96,22 +96,28 @@ private void persistBuffers(JfrChunkWriter chunkWriter) {
 
+    @Uninterruptible(reason = "Epoch must not change while in this method.")
+    private boolean persistBuffer(JfrChunkWriter chunkWriter, JfrBuffer buffer) {
+        if (isFullEnough(buffer) && JfrBufferAccess.acquire(buffer)) {
```
The check for `isFullEnough(...)` could be moved to the caller of this method.

jiekang commented:
Addressed most comments but still going over the constraints for the top pointer. Noticed that the NativeEvent work seems to result in mismatched or corrupted ThreadStart/ThreadEnd events. Continuing to investigate that as well.

jiekang commented:
Okay, issue with native events found and resolved here: https://github.com/native-image-jfr/graal/pull/26

It didn't have anything to do with the chunk rotation mechanism. I'm still verifying the paths that modify the top pointer.

jiekang commented:
@christianhaeubl @rkennke Thanks for your patience. Please take another look when you have the time; I think I addressed everything so far.

rkennke commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceIdLoadBarrier.java:
```diff
@@ -71,8 +71,7 @@ private static void increaseClassCount(boolean epoch) {
 
-    public static int classCount(boolean epoch) {
-        assert VMOperation.isInProgressAtSafepoint();
+    public static long classCount(boolean epoch) {
```
Why have you changed the return type here?

jiekang replied:
My mistake; I've revised it back now. Thanks for catching this!

rkennke reviewed (changes requested):
Looks fairly good to me. Only little comments.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrBufferAccess.java:
```diff
@@ -68,23 +68,24 @@ public static void free(JfrBuffer buffer) {
     }
 
-    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
+    @Uninterruptible(reason = "Prevent safepoints as those could change the top pointer.", callerMustBe = true)
```
I don't think that `callerMustBe = true` is absolutely necessary here (top is just set to a new value, it is not returned)

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrBufferAccess.java:
```diff
@@ -68,23 +68,24 @@ public static void free(JfrBuffer buffer) {
     }
 
-    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
+    @Uninterruptible(reason = "Called from uninterruptible code.", callerMustBe = true)
```
`@Uninterruptible(reason = "We must guarantee that all buffers are in unacquired state when entering a safepoint.", callerMustBe = true)`

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrBufferAccess.java:
```diff
@@ -68,23 +68,24 @@ public static void free(JfrBuffer buffer) {
     }
 
-    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
+    @Uninterruptible(reason = "Called from uninterruptible code.", callerMustBe = true)
```
Same reason as `acquire`

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrBufferAccess.java:
```diff
@@ -111,16 +112,21 @@ public static UnsignedWord getAvailableSize(JfrBuffer buffer) {
     }
 
+    @Uninterruptible(reason = "Prevent safepoints as those could change the top pointer.", callerMustBe = true)
+    public static void increaseTop(JfrBuffer buffer, UnsignedWord delta) {
```
Here we could remove the `callerMustBe = true`.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -115,11 +121,11 @@ public void maybeOpenFile() {
+            fd = getFileSupport().open(filename, RawFileOperationSupport.FileAccessMode.READ_WRITE);
             writeFileHeader();
             // TODO: this should probably also write all live threads
             return true;
```
I don't think that the `IOException` below can happen anymore.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -129,84 +135,98 @@ public boolean openFile(String outputFile) {
+            SignedWord metadataPosition = writeMetadataEvent(metadataDescriptor);
+            patchFileHeader(constantPoolPosition, metadataPosition);
+            getFileSupport().close(fd);
+        } catch (IOException e) {
```
That `IOException` also can't happen anymore.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -129,84 +135,98 @@ public boolean openFile(String outputFile) {
+        fd = WordFactory.nullPointer();
     }
 
     private void writeFileHeader() throws IOException {
```
That `IOException` also can't happen anymore.

christianhaeubl commented on substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java:
```diff
@@ -129,84 +135,98 @@ public boolean openFile(String outputFile) {
 
-    private void patchFileHeader(long constantPoolPosition, long metadataPosition) throws IOException {
-        long chunkSize = file.getFilePointer();
+    public void patchFileHeader(SignedWord constantPoolPosition, SignedWord metadataPosition) throws IOException {
```
That `IOException` also can't happen anymore.

christianhaeubl reviewed (approved):
Only minor comments.

christianhaeubl commented:
Can this PR be merged?

jiekang commented:
@christianhaeubl Hi! yes I can work to merge it in now. I was initially waiting to rebase against upstream for GR-30929.

jiekang committed:
Implement chunk rotation
